File: sdk/typescript/src/extension-describe.ts
import {
  SDK_MIN_COMPOZY_VERSION,
  SDK_NAME,
  type ExtensionDescribeProcess,
} from "./extension-contract.js";
import { normalizeHostMethodList, normalizeStringList } from "./extension-runtime.js";
import type {
  AutonomyMatcher,
  CmdPaletteConfig,
  DescribePayload,
  DescribeHookEvent,
  DescribeProfile,
  DescribeResourcePath,
  ExtensionDefinition,
  ExtensionCommandGroupSpec,
  ExtensionToolRuntimeDescriptor,
  HookMatcher,
} from "./types.js";
import type { RegisteredTool } from "./extension-contract.js";

interface ExtensionDescribeInput {
  definition: ExtensionDefinition;
  tools: ExtensionToolRuntimeDescriptor[];
  commandGroups: ExtensionCommandGroupSpec[];
  watchSourceKinds: string[];
  sdkVersion: string;
}

const UTF8_ENCODER = new TextEncoder();

export function compareUTF8Strings(left: string, right: string): number {
  if (left === right) {
    return 0;
  }
  const leftBytes = UTF8_ENCODER.encode(left);
  const rightBytes = UTF8_ENCODER.encode(right);
  const sharedLength = Math.min(leftBytes.length, rightBytes.length);
  for (let index = 0; index < sharedLength; index += 1) {
    const compared = (leftBytes[index] ?? 0) - (rightBytes[index] ?? 0);
    if (compared !== 0) {
      return compared;
    }
  }
  return leftBytes.length - rightBytes.length;
}

export function defaultExtensionDescribeProcess(): ExtensionDescribeProcess {
  return {
    argv: process.argv,
    stdout: process.stdout,
    setExitCode: code => {
      process.exitCode = code;
    },
  };
}

export function runExtensionDescribeMode(
  describeProcess: ExtensionDescribeProcess,
  describe: () => DescribePayload
): boolean {
  if (describeProcess.argv.at(-1)?.trim() !== "__describe") {
    return false;
  }
  describeProcess.stdout.write(`${JSON.stringify(describe())}\n`);
  describeProcess.setExitCode(0);
  return true;
}

export function cloneExtensionToolDescriptors(
  tools: Iterable<RegisteredTool>
): ExtensionToolRuntimeDescriptor[] {
  return [...tools].map(tool => ({
    ...tool.descriptor,
    ...(tool.descriptor.input_schema === undefined
      ? {}
      : { input_schema: structuredClone(tool.descriptor.input_schema) }),
    ...(tool.descriptor.output_schema === undefined
      ? {}
      : { output_schema: structuredClone(tool.descriptor.output_schema) }),
    capabilities: [...(tool.descriptor.capabilities ?? [])],
    ...(tool.descriptor.command === undefined
      ? {}
      : { command: structuredClone(tool.descriptor.command) }),
  }));
}

export function cmdPaletteViewIDs(config: CmdPaletteConfig | undefined): string[] {
  return (config?.views ?? []).map(view => view.id.trim()).filter(id => id.length > 0);
}

export function buildExtensionDescribePayload(input: ExtensionDescribeInput): DescribePayload {
  const command = input.definition.subprocess?.command.trim() ?? "";
  if (!command) {
    throw new Error("subprocess command is required");
  }
  const viewIDs = cmdPaletteViewIDs(input.definition.resources?.cmd_palette);

  return {
    name: input.definition.name.trim(),
    version: input.definition.version.trim(),
    ...(input.definition.description?.trim()
      ? { description: input.definition.description.trim() }
      : {}),
    provides: normalizeStringList(input.definition.capabilities?.provides),
    permissions: normalizeHostMethodList(input.definition.permissions?.requires),
    requires_env: normalizeStringList(input.definition.requires_env),
    profiles: normalizeDescribeProfiles(input.definition.profiles),
    resources: {
      skills: normalizeDescribeResourcePaths(input.definition.resources?.skills),
      loops: normalizeDescribeResourcePaths(input.definition.resources?.loops),
      agents: normalizeDescribeResourcePaths(input.definition.resources?.agents),
      automation: normalizeDescribeResourcePaths(input.definition.resources?.automation),
      layouts: normalizeDescribeResourcePaths(input.definition.resources?.layouts),
      ...(input.definition.resources?.cmd_palette === undefined
        ? {}
        : { cmd_palette: structuredClone(input.definition.resources.cmd_palette) }),
    },
    subprocess: {
      command,
      args: [...(input.definition.subprocess?.args ?? [])],
      env: { ...input.definition.subprocess?.env },
    },
    ...(input.definition.network_participation === undefined
      ? {}
      : {
          network_participation: {
            required: input.definition.network_participation.required,
            mode: input.definition.network_participation.mode.trim().toLowerCase(),
            channel_scopes: normalizeStringList(
              input.definition.network_participation.channel_scopes
            ),
          },
        }),
    tools: [...input.tools].sort((left, right) => left.handler.localeCompare(right.handler)),
    command_groups: input.commandGroups.map(group => ({ ...group })),
    hook_events: normalizeDescribeHookEvents(input.definition.supported_hook_events),
    watch_source_kinds: input.watchSourceKinds,
    ...(viewIDs.length === 0 ? {} : { cmd_palette_views: viewIDs }),
    sdk: {
      name: SDK_NAME,
      version: input.sdkVersion,
      protocol_version: "1",
      min_compozy_version: SDK_MIN_COMPOZY_VERSION,
    },
  };
}

function normalizeDescribeResourcePaths(
  resources: DescribeResourcePath[] | undefined
): DescribeResourcePath[] {
  const unique = new Map<string, DescribeResourcePath>();
  for (const resource of resources ?? []) {
    const normalized = { path: resource.path.trim(), profile: resource.profile?.trim() };
    unique.set(`${normalized.path}\u0000${normalized.profile ?? ""}`, {
      path: normalized.path,
      ...(normalized.profile ? { profile: normalized.profile } : {}),
    });
  }
  return [...unique.values()].sort(
    (left, right) =>
      left.path.localeCompare(right.path) || (left.profile ?? "").localeCompare(right.profile ?? "")
  );
}

function normalizeDescribeHookEvents(events: DescribeHookEvent[] | undefined): DescribeHookEvent[] {
  const unique = new Map<string, DescribeHookEvent>();
  for (const item of events ?? []) {
    const event = item.event.trim() as DescribeHookEvent["event"];
    const name = normalizeDetectableDescribeOptional(item.name);
    const profile = item.profile?.trim();
    const mode = normalizeDetectableDescribeOptional(item.mode) as DescribeHookEvent["mode"];
    const matcher = item.matcher === undefined ? undefined : cloneDescribeHookMatcher(item.matcher);
    const normalized: DescribeHookEvent = {
      ...(name !== undefined ? { name } : {}),
      event,
      ...(profile ? { profile } : {}),
      ...(mode !== undefined ? { mode } : {}),
      ...(matcher !== undefined && Object.keys(matcher).length > 0 ? { matcher } : {}),
      ...(item.required ? { required: true } : {}),
    };
    const key = describeHookEventKey(normalized);
    if (!unique.has(key)) {
      unique.set(key, normalized);
    }
  }
  return [...unique.values()].sort(
    (left, right) =>
      compareUTF8Strings(left.profile ?? "", right.profile ?? "") ||
      compareUTF8Strings(left.event, right.event) ||
      compareUTF8Strings(left.name ?? "", right.name ?? "") ||
      compareUTF8Strings(describeHookEventKey(left), describeHookEventKey(right))
  );
}

function normalizeDetectableDescribeOptional(value: string | undefined): string | undefined {
  // Exact empty follows Go's value-typed absent form; non-empty whitespace must remain rejectable.
  if (value === undefined || value === "") {
    return undefined;
  }
  const normalized = value.trim();
  return normalized === "" ? value : normalized;
}

type StringKeyOf<T> = {
  [Key in keyof T]-?: Exclude<T[Key], undefined> extends string ? Key : never;
}[keyof T];

type HookMatcherStringKey = StringKeyOf<HookMatcher>;
type AutonomyMatcherStringKey = StringKeyOf<AutonomyMatcher>;

function defineExhaustiveStringKeys<AllKeys extends string>() {
  return <Keys extends readonly AllKeys[]>(
    keys: Keys &
      (Exclude<AllKeys, Keys[number]> extends never
        ? unknown
        : { missing: Exclude<AllKeys, Keys[number]> })
  ): Keys => keys;
}

const HOOK_MATCHER_STRING_KEYS_BEFORE_TOOL_READ_ONLY = [
  "agent_name",
  "agent_type",
  "workspace_id",
  "worktree_id",
  "workspace_root",
  "session_type",
  "sandbox_id",
  "sandbox_backend",
  "sandbox_profile",
  "sync_direction",
  "input_class",
  "acp_event_type",
  "turn_id",
  "tool_id",
  "tool_name",
] as const satisfies readonly HookMatcherStringKey[];

const HOOK_MATCHER_STRING_KEYS_AFTER_TOOL_READ_ONLY = [
  "decision_class",
  "message_role",
  "message_delta_type",
  "channel",
  "surface",
  "kind",
  "direction",
  "work_state",
  "participation_mode",
  "participation_source",
  "compaction_reason",
  "compaction_strategy",
] as const satisfies readonly HookMatcherStringKey[];

const HOOK_MATCHER_STRING_KEYS = defineExhaustiveStringKeys<HookMatcherStringKey>()([
  ...HOOK_MATCHER_STRING_KEYS_BEFORE_TOOL_READ_ONLY,
  ...HOOK_MATCHER_STRING_KEYS_AFTER_TOOL_READ_ONLY,
] as const);

const AUTONOMY_MATCHER_STRING_KEYS = defineExhaustiveStringKeys<AutonomyMatcherStringKey>()([
  "task_id",
  "run_id",
  "loop_run_id",
  "loop_name",
  "node_id",
  "workflow_id",
  "participation_channel",
  "coordinator_session_id",
  "parent_session_id",
  "root_session_id",
  "child_session_id",
  "spawn_role",
  "release_reason",
] as const);

function cloneDescribeHookMatcher(matcher: HookMatcher): HookMatcher {
  const cloned: HookMatcher = {};
  const target = cloned as Record<HookMatcherStringKey, string | undefined>;
  copyTrimmedMatcherStrings(target, matcher, HOOK_MATCHER_STRING_KEYS);
  if (matcher.tool_read_only !== undefined) {
    cloned.tool_read_only = matcher.tool_read_only;
  }
  if (matcher.autonomy !== undefined) {
    cloned.autonomy = cloneDescribeAutonomyMatcher(matcher.autonomy);
  }
  return cloned;
}

function copyTrimmedMatcherStrings(
  target: Record<HookMatcherStringKey, string | undefined>,
  matcher: HookMatcher,
  keys: readonly HookMatcherStringKey[]
): void {
  for (const key of keys) {
    const value = matcher[key]?.trim();
    if (value) {
      target[key] = value;
    }
  }
}

function cloneDescribeAutonomyMatcher(matcher: AutonomyMatcher): AutonomyMatcher {
  const cloned: AutonomyMatcher = {};
  const target = cloned as Record<AutonomyMatcherStringKey, string | undefined>;
  for (const key of AUTONOMY_MATCHER_STRING_KEYS) {
    const value = matcher[key]?.trim();
    if (value) {
      target[key] = value;
    }
  }
  return cloned;
}

function describeHookEventKey(event: DescribeHookEvent): string {
  const parts: string[] = [];
  appendDescribeKeyPart(parts, event.profile ?? "");
  appendDescribeKeyPart(parts, event.event);
  appendDescribeKeyPart(parts, event.name ?? "");
  appendDescribeKeyPart(parts, event.mode ?? "");
  appendDescribeHookMatcherKey(parts, event.matcher);
  parts.push(event.required ? "1" : "0");
  return parts.join("");
}

function appendDescribeHookMatcherKey(parts: string[], matcher: HookMatcher | undefined): void {
  for (const key of HOOK_MATCHER_STRING_KEYS_BEFORE_TOOL_READ_ONLY) {
    appendDescribeKeyPart(parts, matcher?.[key] ?? "");
  }
  parts.push(matcher?.tool_read_only === undefined ? "n" : matcher.tool_read_only ? "t" : "f");
  for (const key of HOOK_MATCHER_STRING_KEYS_AFTER_TOOL_READ_ONLY) {
    appendDescribeKeyPart(parts, matcher?.[key] ?? "");
  }
  if (matcher?.autonomy === undefined) {
    parts.push("n");
    return;
  }
  parts.push("v");
  for (const key of AUTONOMY_MATCHER_STRING_KEYS) {
    appendDescribeKeyPart(parts, matcher.autonomy[key] ?? "");
  }
}

function appendDescribeKeyPart(parts: string[], value: string): void {
  parts.push(`${UTF8_ENCODER.encode(value).length}:${value}`);
}

function normalizeDescribeProfiles(profiles: DescribeProfile[] | undefined): DescribeProfile[] {
  return [...(profiles ?? [])]
    .map(profile => {
      const defaults = profile.defaults ?? {};
      return {
        name: profile.name.trim(),
        ...(profile.color?.trim() ? { color: profile.color.trim() } : {}),
        ...(profile.icon?.trim() ? { icon: profile.icon.trim() } : {}),
        ...(profile.emoji?.trim() ? { emoji: profile.emoji.trim() } : {}),
        defaults: {
          ...(defaults.agent?.trim() ? { agent: defaults.agent.trim() } : {}),
          ...(defaults.provider?.trim() ? { provider: defaults.provider.trim() } : {}),
          ...(defaults.sandbox?.trim() ? { sandbox: defaults.sandbox.trim() } : {}),
        },
        credentials: [...(profile.credentials ?? [])]
          .map(credential => ({
            provider: credential.provider.trim(),
            slot: credential.slot.trim(),
          }))
          .sort(
            (left, right) =>
              left.provider.localeCompare(right.provider) || left.slot.localeCompare(right.slot)
          ),
      };
    })
    .sort((left, right) => left.name.localeCompare(right.name));
}

File: sdk/go/extension_describe.go
package compozysdk

import (
	"encoding/json"
	"io"
	"maps"
	"slices"
	"strings"

	"github.com/compozy/compozy/sdk/go/contracts"
)

const describeArgument = "__describe"

// Describe returns the deterministic build-time contract assembled from SDK registrations.
func (e *Extension) Describe() (contracts.DescribePayload, error) {
	if e == nil {
		return contracts.DescribePayload{}, NewInternalError("extension is required")
	}

	e.mu.RLock()
	defer e.mu.RUnlock()
	if err := e.definition.validate(); err != nil {
		return contracts.DescribePayload{}, err
	}
	if strings.TrimSpace(e.definition.Subprocess.Command) == "" {
		return contracts.DescribePayload{}, NewInvalidParamsError("subprocess command is required", nil)
	}

	tools := make([]contracts.ExtensionToolRuntimeDescriptor, 0, len(e.toolHandlers))
	for _, registered := range e.toolHandlers {
		descriptor := registered.descriptor
		tools = append(tools, contracts.ExtensionToolRuntimeDescriptor{
			Profile:             descriptor.Profile,
			ID:                  contracts.ToolID(descriptor.ID),
			Handler:             descriptor.Handler,
			Description:         descriptor.Description,
			FriendlyVerb:        descriptor.FriendlyVerb,
			Preview:             descriptor.Preview,
			InputSchema:         cloneRawMessage(descriptor.InputSchema),
			OutputSchema:        cloneRawMessage(descriptor.OutputSchema),
			InputSchemaDigest:   descriptor.InputSchemaDigest,
			OutputSchemaDigest:  descriptor.OutputSchemaDigest,
			ReadOnly:            descriptor.ReadOnly,
			Risk:                contracts.RiskClass(descriptor.Risk),
			RequiresInteraction: descriptor.RequiresInteraction,
			Capabilities:        slices.Clone(descriptor.Capabilities),
			Command:             cloneSDKCommandSpec(descriptor.Command),
		})
	}
	slices.SortFunc(tools, func(left, right contracts.ExtensionToolRuntimeDescriptor) int {
		return strings.Compare(left.Handler, right.Handler)
	})

	permissions := make([]string, 0, len(e.definition.Permissions.Requires))
	for _, method := range e.definition.Permissions.Requires {
		permissions = append(permissions, string(method))
	}
	minCompozyVersion := strings.TrimSpace(e.definition.MinCompozyVersion)
	if minCompozyVersion == "" {
		minCompozyVersion = MinCompozyVersion
	}

	return contracts.DescribePayload{
		Name:        strings.TrimSpace(e.definition.Name),
		Version:     strings.TrimSpace(e.definition.Version),
		Description: strings.TrimSpace(e.definition.Description),
		Provides:    requiredStringList(e.definition.Capabilities.Provides),
		Permissions: requiredStringList(permissions),
		RequiresEnv: normalizeStrings(e.definition.RequiresEnv),
		Profiles:    normalizeDescribeProfiles(e.definition.Profiles),
		Resources: contracts.DescribeResources{
			Skills:     normalizeDescribeResourcePaths(e.definition.Resources.Skills),
			Loops:      normalizeDescribeResourcePaths(e.definition.Resources.Loops),
			Agents:     normalizeDescribeResourcePaths(e.definition.Resources.Agents),
			Automation: normalizeDescribeResourcePaths(e.definition.Resources.Automation),
			Layouts:    normalizeDescribeResourcePaths(e.definition.Resources.Layouts),
			CmdPalette: cloneCmdPaletteConfig(e.definition.Resources.CmdPalette),
		},
		Subprocess: contracts.DescribeSubprocess{
			Command: strings.TrimSpace(e.definition.Subprocess.Command),
			Args:    slices.Clone(e.definition.Subprocess.Args),
			Env:     cloneStringMap(e.definition.Subprocess.Env),
		},
		NetworkParticipation: normalizeDescribeNetworkParticipation(e.definition.NetworkParticipation),
		Tools:                tools,
		CommandGroups:        e.commandGroupsLocked(),
		HookEvents:           normalizeDescribeHookEvents(e.definition.SupportedHookEvents),
		WatchSourceKinds:     e.watchSourceKindsLocked(),
		CmdPaletteViews:      cmdPaletteViewIDs(e.definition.Resources.CmdPalette),
		SDK: contracts.DescribeSDKInfo{
			Name:              SDKName,
			Version:           e.sdkVersion,
			ProtocolVersion:   ProtocolVersion,
			MinCompozyVersion: minCompozyVersion,
		},
	}, nil
}

func cmdPaletteViewIDs(config contracts.CmdPaletteConfig) []string {
	ids := make([]string, 0, len(config.Views))
	for _, view := range config.Views {
		id := strings.TrimSpace(view.ID)
		if id == "" {
			continue
		}
		ids = append(ids, id)
	}
	return ids
}

func cloneCmdPaletteConfig(value contracts.CmdPaletteConfig) contracts.CmdPaletteConfig {
	cloned := contracts.CmdPaletteConfig{
		Commands: slices.Clone(value.Commands),
		Views:    slices.Clone(value.Views),
	}
	for index := range cloned.Commands {
		command := &cloned.Commands[index]
		command.Keywords = slices.Clone(value.Commands[index].Keywords)
		command.Arguments = slices.Clone(value.Commands[index].Arguments)
		for argumentIndex := range command.Arguments {
			command.Arguments[argumentIndex].Options = slices.Clone(
				value.Commands[index].Arguments[argumentIndex].Options,
			)
		}
		command.Action.Args = cloneJSONMap(value.Commands[index].Action.Args)
		if value.Commands[index].Confirmation != nil {
			confirmation := *value.Commands[index].Confirmation
			command.Confirmation = &confirmation
		}
		if value.Commands[index].Execution != nil {
			execution := *value.Commands[index].Execution
			execution.SingleFlight = cloneOptionalBool(value.Commands[index].Execution.SingleFlight)
			execution.RetrySafe = cloneOptionalBool(value.Commands[index].Execution.RetrySafe)
			command.Execution = &execution
		}
	}
	for index := range cloned.Views {
		if value.Views[index].Source != nil {
			source := *value.Views[index].Source
			cloned.Views[index].Source = &source
		}
	}
	return cloned
}

func cloneJSONMap(values map[string]any) map[string]any {
	if values == nil {
		return nil
	}
	cloned := make(map[string]any, len(values))
	for key, item := range values {
		cloned[key] = cloneJSONValue(item)
	}
	return cloned
}

func cloneJSONValue(value any) any {
	switch typed := value.(type) {
	case nil:
		return nil
	case json.RawMessage:
		return slices.Clone(typed)
	case map[string]any:
		return cloneJSONMap(typed)
	case map[string]string:
		return maps.Clone(typed)
	case []any:
		cloned := make([]any, len(typed))
		for index, item := range typed {
			cloned[index] = cloneJSONValue(item)
		}
		return cloned
	case []string:
		return slices.Clone(typed)
	case []map[string]any:
		cloned := make([]map[string]any, len(typed))
		for index, item := range typed {
			cloned[index] = cloneJSONMap(item)
		}
		return cloned
	case []map[string]string:
		cloned := make([]map[string]string, len(typed))
		for index, item := range typed {
			cloned[index] = maps.Clone(item)
		}
		return cloned
	default:
		return typed
	}
}

func cloneOptionalBool(value *bool) *bool {
	if value == nil {
		return nil
	}
	cloned := *value
	return &cloned
}

func normalizeDescribeNetworkParticipation(
	value *NetworkParticipationRequirement,
) *contracts.DescribeNetworkParticipation {
	if value == nil {
		return nil
	}
	return &contracts.DescribeNetworkParticipation{
		Required: value.Required, Mode: strings.TrimSpace(value.Mode),
		ChannelScopes: normalizeStrings(value.ChannelScopes),
	}
}

func (e *Extension) writeDescribe(output io.Writer) error {
	payload, err := e.Describe()
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(output)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(payload); err != nil {
		return NewInternalError("encode describe payload: " + err.Error())
	}
	return nil
}

func describeModeRequested(args []string) bool {
	return len(args) > 1 && strings.TrimSpace(args[len(args)-1]) == describeArgument
}

func requiredStringList(values []string) []string {
	result := normalizeStrings(values)
	if result == nil {
		return []string{}
	}
	return result
}

func cloneStringMap(values map[string]string) map[string]string {
	if len(values) == 0 {
		return nil
	}
	return maps.Clone(values)
}
